### PR TITLE
Enable PCIe SRIS for Tofino

### PR DESCRIPTION
**Pull request by arjenroodselaar**
This diff enables PCIe Separate Ref clock Independent Spread (SRIS) for Tofino, allowing us to enable spread spectrum clocking for Scrimlet hosts as determined to be necessary during EMC testing. The diff leaves the old clock config (with the Tofino ref clock tracking the incoming SP3 generated clock for common ref clock) in case someone needs that in the near future.

### CODE REVIEW

File: drv/sidecar-mainboard-controller/src/tofino2.rs [hunk 1]
```diff
@@ -493,22 +494,135 @@ pub enum DirectBarSegment {
...
+    #[derive(Copy, Clone, PartialEq, Eq, From, FromPrimitive, AsBytes, FromBytes)]
+    #[repr(C)]
+    pub struct SoftwareReset(u32);
+    pub pcie_phy, set_pcie_phy: 0;
+    pub pcie_ctrl, set_pcie_ctrl: 2;
+    pub pcie_app, set_pcie_app: 3;
+    pub pcie_lanes, set_pcie_lanes: 7, 4;
+}
+
+impl From<SoftwareReset> for u32 {
```
[the remark this answers is not in the document]
Comment (mkeeter) [reply drv/sidecar-mainboard-controller/src/tofino2.rs:541]: Also, `FromPrimitive` is probably unnecessary (unless we need it for Idol encoding or something)

File: drv/sidecar-mainboard-controller/src/tofino2.rs [hunk 2]
```diff
@@ -493,22 +494,135 @@ pub enum DirectBarSegment {
...
+    fn from(r: SoftwareReset) -> Self {
+        r.0
+    }
+}
+
+bitfield! {
+    #[derive(Copy, Clone, PartialEq, Eq, From, FromPrimitive, AsBytes, FromBytes)]
```
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/tofino2.rs:548]: Same here (add `Into`, remove `FromPrimitive`)

File: drv/sidecar-mainboard-controller/src/tofino2.rs [hunk 3]
```diff
@@ -493,22 +494,135 @@ pub enum DirectBarSegment {
...
+}
+
+impl From<PciePhyLaneControl> for u16 {
+    fn from(r: PciePhyLaneControl) -> Self {
+        r.0
+    }
+}
+
+bitfield! {
+    #[derive(Copy, Clone, PartialEq, Eq, From, FromPrimitive, AsBytes, FromBytes)]
```
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/tofino2.rs:569]: Same here

File: drv/sidecar-mainboard-controller/src/tofino2.rs [hunk 4]
```diff
@@ -493,22 +494,135 @@ pub enum DirectBarSegment {
...
+}
+
+impl From<PciePhyLaneControlPair> for u32 {
+    fn from(r: PciePhyLaneControlPair) -> Self {
+        r.0
+    }
+}
+
+bitfield! {
+    #[derive(Copy, Clone, PartialEq, Eq, From, FromPrimitive, AsBytes, FromBytes)]
```
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/tofino2.rs:583]: etc

File: drv/sidecar-mainboard-controller/src/tofino2.rs [hunk 5]
```diff
@@ -493,22 +494,135 @@ pub enum DirectBarSegment {
...
+    #[repr(C)]
+    pub struct PciePhyLaneStatus(u16);
+    pub elastic_buffer_pointer, _: 7, 0;
+    pub aligned, _: 8;
+    pub pipe_data_bus_width, _: 11, 10;
+    pub pipe_max_p_clk, _: 13, 12;
+}
+
+bitfield! {
+    #[derive(Copy, Clone, PartialEq, Eq, From, FromPrimitive, AsBytes, FromBytes)]
```
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/tofino2.rs:593]: etc

File: drv/sidecar-mainboard-controller/src/tofino2.rs [hunk 6]
```diff
@@ -493,22 +494,135 @@ pub enum DirectBarSegment {
...
+}
+
+impl From<PciePhyLaneStatusPair> for u32 {
+    fn from(r: PciePhyLaneStatusPair) -> Self {
+        r.0
+    }
+}
+
+bitfield! {
+    #[derive(Copy, Clone, PartialEq, Eq, From, FromPrimitive, AsBytes, FromBytes)]
```
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/tofino2.rs:607]: etc

File: drv/sidecar-mainboard-controller/src/tofino2.rs [hunk 7]
```diff
@@ -618,7 +732,7 @@ impl DebugPort {
         &self,
         segment: DirectBarSegment,
         offset: impl Into<u32> + Copy,
-        value: u32,
+        value: impl Into<u32> + AsBytes,
```
Comment (mkeeter) [inline drv/sidecar-mainboard-controller/src/tofino2.rs:735]:
```suggestion
        value: impl AsBytes,
```
(we don't actually use the `Into<u32>` bound)

File: drv/sidecar-seq-server/src/clock_generator_payload_b.rs [hunk 8]
```diff
@@ -782,7 +782,7 @@ pub fn idt8a3xxxx_payload<E>(
             0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
             0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
             0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-            0x00, 0x00, 0x00, 0x01, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
```
Comment (mkeeter) [inline drv/sidecar-seq-server/src/clock_generator_payload_b.rs:785]: Is it expected that this value changed, but all of the comments above stayed the same?

**Review by mkeeter — APPROVED** [review]
A few stylistic suggestions, but generally seems fine

File: drv/sidecar-mainboard-controller/src/tofino2.rs (hunk 7 above)
Comment (arjenroodselaar) [reply drv/sidecar-mainboard-controller/src/tofino2.rs:735]: It does now, taking one of the u32 NewTypes and pulling out the bytes.

File: drv/sidecar-seq-server/src/clock_generator_payload_b.rs (hunk 8 above)
Comment (arjenroodselaar) [reply drv/sidecar-seq-server/src/clock_generator_payload_b.rs:785]: Yes. This change indicates that particular DPLL does not track input[1] anymore. But the parser used to generate this does not seem to pick that up.

**Pull request merged**